text/3696-unsized-params-in-traits.md

```diff
@@ -0,0 +1,168 @@
+
+While both of the above _would_ work without the lint, you cant actually implement it
+```rust
+impl Foo for i32 {
+    fn bar(bytes: [u8]) {}
+}
+```
+Produces:
+```
+error: The Size value for `[u8]` is not known at compile time
+```
+
+So in all: this rfc is to prevent confusion
+
+Now, if you do notice, in the [summary], i did say `required methods`; provided methods are `Sized`-checked
+```rust
+trait Foo {
+    fn bar(self) {
+
+    }
+}
+```
+Produces:
+```
+error: the size value for `Self` is not know at compile time
+```
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There is one feature this may clash with, and it is the feature for `unsized fn params`, the lint should be disabled when this feature is enabled; if that is possible, however if it may cause confusion to the user.
+
+Here is the first example, but it clashes with the `unsized fn params` feature
+```rust
+#![feature(unsized_fn_params)]
+// implicit #![deny(unsized_params_in_trait)]
+#![allow(internal_features)]
+
+trait Foo {
+    fn bar(bytes: [u8]);
+}
+```
+
+The above code fails, while it shouldn't due to the feature
+
+However, it is internal so it is semi-ok
+
+this feature will be very simple to implement (i hope), just port the `Sized`-checking logic from provided methods to required methods, if that is possible (also maybe from regular functions) and throw an error/warning/nothing depending on the lint level.
+
+here is some rust-pseudo code:
+```rust
+if !req_function.
+    params
+    .filter(|param| !param.is_sized())
+    .is_empty() {
+        match lint_level {
+            Allow => (),
+            Warn => warn("..."),
+            Deny | Forbid => error("...")
+        }
+    }
+```
+replace the `...` with: 
+```
+The size value for {param_type} is not know at compile time
+# ...
+This was previously accepted by the compiler, but it will become a hard error in the future!
+# if it was the default value of 'deny' the next line would be added
+`#[deny(unsized_param_in_trait)]` on by default
+```
+Obviously the above code isnt actually correct as it doesnt check _which_ param is unsized, it just checks if there is, (you can probably loop over the 'filter' object, and make an individual error for each one)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This could cause breaking changes, however the lint gives time for migration
+- This could be intended for `dyn` compatibility (see [rationale-and-alternatives] for a way to fix this)
+This drawback is about receivers, take this example
+```rust
+trait Bar {
+    fn foo(self);
+}
+```
+if `self` was sized checked here, and the value _should_ be consumed, then this code would be impossible without a `Self: Sized` bound, but as you know, adding that bound removes `dyn-Compatibility`
+```rust
+let x: &dyn Bar = &10;
+```
+Produces:
+```
+Bar is not dyn compatible
+# ...
+```
+
+# Rationale and alternatives
```

Comment:
An alternative could be to generate `where Self: Sized` bounds from the signature where applicable and thus never actually forbid it.

Reply:
This is what the Reference (inaccurately) says currently happens, for context.

<details> <summary>source</summary>

https://doc.rust-lang.org/reference/items/traits.html#object-safety

> [...] A trait is object safe if it has the following qualities [...] All associated functions must either be dispatchable from a trait object or be explicitly non-dispatchable: [...] Dispatchable functions must: [...] Not have a where Self: Sized bound **(receiver type of Self (i.e. self) implies this)**. [...] Explicitly non-dispatchable functions require: Have a where Self: Sized bound **(receiver type of Self (i.e. self) implies this)**.

(emphasis mine)

</details>

Reply:
This would, for the record, break the dyn compatibility of `FnOnce`.

Reply:
So does this RFC. And we'd have to revert this RFC whenever we figure out generally allowing folk to call owned methods on `Box<dyn TheirTrait>`